### PR TITLE
Add annotations for Signedness checker

### PR DESCRIPTION
**Pull request by markro49**
Before pulling master into this branch, merge https://github.com/typetools/commons-bcel/pull/20 .

### CODE REVIEW

**Comment by mernst** [comment]
This pull request includes multiple different changes (all good, but still confusing to mix).

Could you separate some of them out?
For example, I agree that tabs should not be used in `pom.xml`, but that should be a separate commit to reduce the size of the diffs.  (I made such a commit to master.)

The documentation about using a locally-built Checker Framework is also useful, but again should be separated out.

Likewise for updating the Checker Framework version number (which can now be 3.23.0).
(I made a pull request.)

Thanks.

**Comment by mernst** [comment]
@markro49 A ping on this pull request.

File: src/main/java/org/apache/bcel/classfile/ConstantUtf8.java [hunk 1]
```diff
@@ -62,6 +62,7 @@ private static class Cache {
 
         private static final boolean BCEL_STATISTICS = Boolean.getBoolean(SYS_PROP_STATISTICS);
         private static final int MAX_ENTRIES = Integer.getInteger(SYS_PROP_CACHE_MAX_ENTRIES, 0).intValue();
+        @SuppressWarnings("cast.unsafe")
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/ConstantUtf8.java]: Can you prefix this by a checker name, so that readers know what error is being suppressed?

File: pom.xml [hunk 2]
```diff
@@ -27,13 +27,13 @@
...
     <artifactId>commons-parent</artifactId>
-    <version>50</version>
+    <version>53</version>
   </parent>
 
   <groupId>org.apache.bcel</groupId>
   <artifactId>bcel</artifactId>
   <packaging>jar</packaging>
-  <version>6.5.0</version>
+  <version>6.6.0-SNAPSHOT</version>
```
Comment (mernst) [inline pom.xml:36]: This version number contains "-SNAPSHOT", but just below on line 51 it does not.  Is that inconsistency intentional?

File: .github/workflows/maven.yml [hunk 3]
```diff
@@ -15,15 +15,15 @@
 
 name: Java CI
 
-on: [push]
+on: [push, pull_request]
```
Comment (mernst) [inline .github/workflows/maven.yml:18]: Should this change (and the other in this file, and possibly a few of those in `pom.xml`) be pull-requested into codespec/commons-bcel rather than typetools/bcel?

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 4]
```diff
@@ -246,7 +249,7 @@ protected void adjustSize() {
      * @param str String to search for
      * @return index on success, -1 otherwise
      */
-    public int lookupString( final String str ) {
+    public @NonNegative int lookupString( final String str ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: This method can return -1, so I think it should be annotated as `@GTENegativeOne`.

Why does this method type-check, given that the return type is inconsistent with the `return` statement?

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 5]
```diff
@@ -258,8 +261,8 @@ public int lookupString( final String str ) {
      * @param str String to add
      * @return index of entry
      */
-    public int addString( final String str ) {
-        int ret;
+    public @NonNegative int addString( final String str ) {
+        @NonNegative int ret;
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: `ret` can be -1.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 6]
```diff
@@ -283,14 +286,14 @@ public int addString( final String str ) {
      * @param str String to search for
      * @return index on success, -1 otherwise
      */
-    public int lookupClass( final String str ) {
+    public @NonNegative int lookupClass( final String str ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: This method can return -1, so I think it should be annotated as `@GTENegativeOne`.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 7]
```diff
@@ -283,14 +286,14 @@ public int addString( final String str ) {
...
         final Index index = classTable.get(str.replace('.', '/'));
         return (index != null) ? index.index : -1;
     }
 
 
-    private int addClass_( final String clazz ) {
-        int ret;
+    private @NonNegative int addClass_( final String clazz ) {
+        @NonNegative int ret;
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: `ret` can be -1.

You shoudn't have to annotate local variables, unless you choose to do so as a stylistic choice.
(Adding the annotations may make merge conflicts more likely in the future.)

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 8]
```diff
@@ -345,7 +348,7 @@ public int addArrayClass( final ArrayType type ) {
      * @param n integer number to look for
      * @return index on success, -1 otherwise
      */
-    public int lookupInteger( final int n ) {
+    public @NonNegative int lookupInteger( final int n ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: This can return -1.  Also fix local variables that are set to uses of it.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 9]
```diff
@@ -382,7 +385,7 @@ public int addInteger( final int n ) {
      * @param n Float number to look for
      * @return index on success, -1 otherwise
      */
-    public int lookupFloat( final float n ) {
+    public @NonNegative int lookupFloat( final float n ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: Can return -1.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 10]
```diff
@@ -422,7 +425,7 @@ public int addFloat( final float n ) {
      * @param n Utf8 string to look for
      * @return index on success, -1 otherwise
      */
-    public int lookupUtf8( final String n ) {
+    public @NonNegative int lookupUtf8( final String n ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: Can return -1.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 11]
```diff
@@ -455,7 +458,7 @@ public int addUtf8( final String n ) {
      * @param n Long number to look for
      * @return index on success, -1 otherwise
      */
-    public int lookupLong( final long n ) {
+    public @NonNegative int lookupLong( final long n ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: Can return -1.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 12]
```diff
@@ -493,7 +496,7 @@ public int addLong( final long n ) {
      * @param n Double number to look for
      * @return index on success, -1 otherwise
      */
-    public int lookupDouble( final double n ) {
+    public @NonNegative int lookupDouble( final double n ) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: Can return -1.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 13]
```diff
@@ -549,10 +552,10 @@ public int lookupNameAndType( final String name, final String signature ) {
      * @param signature signature string to add
      * @return index of entry
      */
-    public int addNameAndType( final String name, final String signature ) {
-        int ret;
-        int name_index;
-        int signature_index;
+    public @NonNegative int addNameAndType( final String name, final String signature ) {
+        @NonNegative int ret;
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: Can be -1.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java [hunk 14]
```diff
@@ -549,10 +552,10 @@ public int lookupNameAndType( final String name, final String signature ) {
...
+        @NonNegative int name_index;
+        @NonNegative int signature_index;
         if ((ret = lookupNameAndType(name, signature)) != -1) {
             return ret; // Already in CP
         }
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: `lookupMethodref` (below) can be annotated as `@GTENegativeOne`.
Likewise `lookupInterfaceRef` and their uses.

File: src/main/java/org/apache/bcel/classfile/Utility.java [hunk 15]
```diff
@@ -1584,13 +1587,15 @@ public int read() throws IOException {
         }
 
 
+/* unused
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/Utility.java]: Even though it is unused, I would retain it (even if you have to suppress warnings within it), to reduce differences from upstream.

File: src/main/java/org/apache/bcel/classfile/Utility.java [hunk 16]
```diff
@@ -1604,29 +1609,27 @@ public JavaWriter(final Writer out) {
...
+                    /* b is guaranteed in range x'30' to x'ff' */
+                    final String tmp = Integer.toHexString(b);
+                    out.write(tmp.charAt(0));
+                    out.write(tmp.charAt(1));
                 }
             }
         }
 
 
+/* unused
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/Utility.java]: Even though it is unused, I would retain it (even if you have to suppress warnings within it), to reduce differences from upstream.

File: src/main/java/org/apache/bcel/generic/BasicType.java [hunk 17]
```diff
@@ -80,6 +80,9 @@ public int hashCode() {
      */
     @Override
     public boolean equals( final @Nullable Object _type ) {
-        return (_type instanceof BasicType) ? ((BasicType) _type).getType() == this.getType() : false;
+        if (_type instanceof BasicType) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/BasicType.java:85]: Why is this change necessary?

File: src/main/java/org/apache/bcel/classfile/LineNumber.java [hunk 18]
```diff
@@ -95,31 +102,31 @@ public void dump( final DataOutputStream file ) throws IOException {
     /**
      * @return Corresponding source line
      */
-    public int getLineNumber() {
-        return 0xffff & lineNumber;
+    public @NonNegative int getLineNumber() {
+        return lineNumber & 0xffff;
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/LineNumber.java:108]: Why is this change necessary?

File: src/main/java/org/apache/bcel/classfile/LineNumber.java [hunk 19]
```diff
@@ -95,31 +102,31 @@ public void dump( final DataOutputStream file ) throws IOException {
...
     }
 
 
     /**
      * @return PC in code
      */
-    public int getStartPC() {
-        return  0xffff & startPc;
+    public @NonNegative int getStartPC() {
+        return  startPc & 0xffff;
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/LineNumber.java:117]: Why is this change necessary?

File: src/main/java/org/apache/bcel/generic/MethodGen.java [hunk 20]
```diff
@@ -826,11 +826,15 @@ public void removeNOPs() {
     /**
      * Set maximum number of local variables.
      */
-    public void setMaxLocals( final int m ) {
+    public void setMaxLocals( final @NonNegative int m ) {
         maxLocals = m;
     }
 
 
+    /**
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/MethodGen.java]: To reduce diffs against upstream, and to simplify merging, please don't add or reformat Javadoc.  (Same comment applies below.)

File: src/main/java/org/apache/bcel/generic/ObjectType.java [hunk 21]
```diff
@@ -75,9 +75,10 @@ public int hashCode() {
      */
     @Override
     public boolean equals( final @Nullable Object type ) {
-        return (type instanceof ObjectType)
-                ? ((ObjectType) type).className.equals(className)
-                : false;
+        if (type instanceof ObjectType) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/ObjectType.java:80]: Why is this change necessary?

File: src/main/java/org/apache/bcel/classfile/LineNumber.java [hunk 22]
```diff
@@ -86,6 +92,7 @@ public void accept( final Visitor v ) {
      * @param file Output file stream
      * @throws IOException if an I/O Exception occurs in writeShort
      */
+    @SuppressWarnings("argument") // writeShort only writes low 16 bits
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/LineNumber.java]: Please qualify with a checker name.

File: src/main/java/org/apache/bcel/generic/Instruction.java [hunk 23]
```diff
@@ -577,7 +577,10 @@ public static void setComparator( final InstructionComparator c ) {
      */
     @Override
     public boolean equals( final @Nullable Object that ) {
-        return (that instanceof Instruction) ? cmp.equals(this, (Instruction) that) : false;
+        if (that instanceof Instruction) {
```
Comment (mernst) [inline src/main/java/org/apache/bcel/generic/Instruction.java:585]: Why is this change necessary?

File: src/main/java/org/apache/bcel/verifier/statics/Pass2Verifier.java [hunk 24]
```diff
@@ -1465,16 +1467,7 @@ private static boolean validClassMethodName(final String name) {
...
-            return false;
-        }
-
-        for (int i=1; i<name.length(); i++) {
-            if (!Character.isJavaIdentifierPart(name.charAt(i))) {
-                return false;
-            }
-        }
-        return true;
+        return validJavaIdentifier(name);
```
Comment (mernst) [inline src/main/java/org/apache/bcel/verifier/statics/Pass2Verifier.java]: Why change this implementation?

File: src/main/java/org/apache/bcel/classfile/Utility.java [hunk 25]
```diff
@@ -1604,29 +1609,27 @@ public JavaWriter(final Writer out) {
         }
 
 
+        @SuppressWarnings("argument") // b is guaranteed in range x'00' to x'ff'
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/Utility.java]: Please add a checker name.

File: src/main/java/org/apache/bcel/classfile/Utility.java [hunk 26]
```diff
@@ -1604,29 +1609,27 @@ public JavaWriter(final Writer out) {
...
                 } else { // Normal escape
-                    final char[] tmp = Integer.toHexString(b).toCharArray();
-                    if (tmp.length == 1) {
-                        out.write('0');
-                        out.write(tmp[0]);
-                    } else {
-                        out.write(tmp[0]);
-                        out.write(tmp[1]);
-                    }
+                    /* b is guaranteed in range x'30' to x'ff' */
```
Comment (mernst) [inline src/main/java/org/apache/bcel/classfile/Utility.java:1623]: Is this and other changes that you have made an improvement that you hope to upstream?
That would be a reason to retain some of the changes to body methods that I noticed.

**Review by mernst — CHANGES_REQUESTED** [review]
Here are some comments and questions.

File: pom.xml (hunk 2 above)
Comment (markro49) [reply pom.xml:36]: As we discussed, this is intentional.  Release version versus interim version.

File: .github/workflows/maven.yml (hunk 3 above)
Comment (markro49) [reply .github/workflows/maven.yml:18]: good idea, I will look into it.

**Review by mernst — COMMENTED** [review]
`LocalVariableInfo.java` and `LocalVariablesInfo.java` do not seem to be annotated.  I expect to see annotations on these `int` varibales:

```java
    public void add(final String name, final int startPc, final int length, final Type type)
```

In `ClassDumper.java`:
    private int access_flags; // Access rights of parsed class

In `AccessFlags.java`:
```java
    public final int getAccessFlags() {
    public final int getModifiers() {
    public final void setAccessFlags(final int access_flags) {
    public final void setModifiers(final int access_flags) {
    private void setFlag(final int flag, final boolean set) {
```

In `ClassParser.java`:
```java
    private int accessFlags; // Access rights of parsed class
```
There are many other unannotated occurrences of `access_flags`, `accessFlags`, `*AccessFlags`, `flags`, and `*Flags`.  I mention these because in `MethodParameter.java` I see the following:

```java
    private @Unsigned int accessFlags;
    public @Unsigned int getAccessFlags() {
```

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java (hunk 5 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: Ret is always positive as if we can't find string we add it.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java (hunk 7 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: I removed annotation of locals.  Ret can not be -1 - if we can't find the class we add it.

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java (hunk 4 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: changed to GTENegativeOne

File: src/main/java/org/apache/bcel/generic/ConstantPoolGen.java (hunk 6 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/ConstantPoolGen.java]: changed to GTENegativeOne

File: src/main/java/org/apache/bcel/generic/BasicType.java (hunk 17 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/BasicType.java:85]: I had to change to avoid following warning:
 [WARNING] /homes/gws/markro/typetools.bcel.master/src/main/java/org/apache/bcel/generic/BasicType.java:[83,8] [[fbc, nullness, a
llcheckers]:contracts.conditional.postcondition] conditional postcondition is not satisfied when equals returns true.
>   found   : _type is @Nullable
>   required: _type is @NonNull

File: src/main/java/org/apache/bcel/generic/ObjectType.java (hunk 21 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/ObjectType.java:80]: same problem as generic/BasicType.java

File: src/main/java/org/apache/bcel/generic/Instruction.java (hunk 23 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/generic/Instruction.java:585]: same problem as generic/BasicType

File: src/main/java/org/apache/bcel/verifier/statics/Pass2Verifier.java (hunk 24 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/verifier/statics/Pass2Verifier.java]: this is a bug fix that is already in upstream apache/commons-bcel.  I will remove it.

File: src/main/java/org/apache/bcel/classfile/Utility.java (hunk 26 above)
Comment (markro49) [reply src/main/java/org/apache/bcel/classfile/Utility.java:1623]: Yes - getting it to check without signedness warnings was easier with these changes and they also fixed a bug.

**Comment by mernst** [comment]
@markro49 Please assign me to this pull request when you have addressed all comments.  I will pull master in and resolve the merge conflicts.  But before doing that, I will merge https://github.com/typetools/commons-bcel/pull/20 .

**Comment by markro49** [comment]
I have addressed everything but the request to review annotations for accessFlags and flags.  can we postpone and make into a new pull request after this one (#15) and #20 are merged?

**Review by mernst — APPROVED** [review]
Thanks for these changes.  You can squash-and-merge anytime you like.
